### PR TITLE
Allow rectangular validation

### PR DESCRIPTION
**Pull request by CySlider**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced dataset image handling to support non-square resolutions.

### 📊 Key Changes
- `imgsz` parameter now accepts a tuple for width and height, allowing different dimensions for image width and height.
- Image processing adjusted to maintain correct aspect ratios based on the new `imgsz` tuple format.
- All references to image size `imgsz` in various methods updated to accommodate tuple values and rectilinear resizing.
- Buffer allocation modified to support non-square image dimensions for mosaics and other augmentation techniques.
- Functions and methods throughout the codebase updated to use the new `imgsz` format.

### 🎯 Purpose & Impact
- **Purpose**: To provide more flexibility in handling images of different dimensions, optimizing the model training for various input shapes.
- **Impact**: Users can now train models with images that aren't constrained to square sizes, improving the versatility and applicability of the models in different scenarios. This should lead to better model performance when dealing with datasets comprised of naturally rectilinear images.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@CySlider thank you for the PR and the detailed summary of your changes! 🌟

Your efforts to enable non-square resolution support in the image transformations are a valuable contribution to the flexibility and generalization capabilities of the model. This is particularly beneficial for datasets with non-standard aspect ratios. 🖼️📐

We'll review the key changes you've made, especially the adjustments to the mosaic augmentation methods and the `LetterBox` transform, to ensure compatibility and performance. Your attention to detail in modifying the `check_cache_ram` and updating image size references is appreciated. 🛠️

Please make sure you've signed the CLA as per the contribution guidelines. Once that's done, we can proceed with integrating your PR into the repository. 📝✅

Looking forward to further enhancing the YOLOv8 experience together!

**Comment by CySlider** [comment]
recheck

**Comment by glenn-jocher** [comment]
@CySlider could you please clarify what you would like to be rechecked? If you're referring to a specific issue or pull request, additional context would be helpful. 🤔

**Comment by glenn-jocher** [comment]
@CySlider hi, sorry, there seems to be a misunderstanding. Rectangular validation is the default for Pytorch models for Detect, Segment, Pose and OBB tasks, so i.e. `yolo val` commands will by default use rectangular batches of the minimum viable size with the validation set presorted by aspect ratio. 

No extra commands or code changes are required.

**Comment by CySlider** [comment]
The "recheck" was for your CLA bot, to recheck my agreement status.

**Comment by CySlider** [comment]
This commit fixes this issue:  [val can't validate exported model if not square resolution was used #7797 ](https://github.com/ultralytics/ultralytics/issues/7797#issuecomment-1913156090)

In short, a model that was exported to engine with dynamic = false could not be validated with rectangular images definition.

**Comment by glenn-jocher** [comment]
@CySlider ok understood. Please review CI failures for this PR in https://github.com/ultralytics/ultralytics/actions/runs/7673573644/job/20916460261

**Comment by CySlider** [comment]
While this seems to work fine for engine, it does even break pt format eval as the image sizes get distorted in that case. So I close this pull request completely now.

**Comment by glenn-jocher** [comment]
@CySlider thank you for your contribution and for detailing the changes in your PR. The ability to handle non-square resolutions is indeed a valuable enhancement for users working with diverse datasets.

Regarding rectangular validation, YOLOv8 already supports this feature. You can enable rectangular inference by setting the `rect` argument to `True` during validation. This allows for more efficient validation by minimizing padding when processing images of varying aspect ratios.

For example, to validate a model with rectangular inference, you can use the following command:

```bash
yolo detect val model=yolov8n.pt rect=True
```

Or in Python:

```python
from ultralytics import YOLO

# Load a model
model = YOLO('yolov8n.pt')

# Validate the model with rectangular inference
metrics = model.val(rect=True)
```

I hope this helps! If you have any further questions or suggestions, please feel free to share. 😊

_I have read the CLA Document and I hereby sign the CLA_
